### ISSUE TITLE
Find definition returns error

### ISSUE THREAD

**Issue by pmatos** [greghendershott/racket-mode#288]
For some reason I keep getting when I do M-. on a function name:
```
racket--repl-command: Process racket-command not running
```

This is with emacs 25.2.1 and racket 6.11.

My init file has:
```
;; Racket Mode
(autoload 'racket-mode "racket-mode" nil t)
(add-to-list 'auto-mode-alist '("\\.rkt\\'" . racket-mode))
(setq racket-racket-program "/home/pmatos/installs/racket-6.11/bin/racket")
(setq racket-raco-program "/home/pmatos/installs/racket-6.11/bin/raco")

; Run with C-c r
(add-hook 'racket-mode-hook
          (lambda ()
            (define-key racket-mode-map (kbd "C-c r") 'racket-run)))

; Enable special racket unicode symbols
(add-hook 'racket-mode-hook      #'racket-unicode-input-method-enable)
(add-hook 'racket-repl-mode-hook #'racket-unicode-input-method-enable)
```

**Comment by greghendershott** [comment]
What happens if you try the equivalent of <kbd>CTRL+ALT+DELETE</kbd> -- enter `,exit` at the Racket REPL prompt, then start it again -- does it work OK then or does the problem persist?

**Comment by pmatos** [comment]
Errr, not sure I understand what `CTRL+ALT+DELETE` equivalent is. Is this an emacs command?

**Comment by greghendershott** [comment]
Sorry, I was joking at myself, because I was basically giving you the classic tech support suggestion to try turning it off and back on.

I meant this part:

1. At a Racket REPL prompt, enter `,exit`.
2. You should see a racket-exited message but the Racket REPL buffer window in Emacs will remain.
3. Then do something to restart it however you usually do (such as <kbd>C-c C-c</kbd> or `M-x racket-repl` or whatever).

Only a very few times have I ever seen the message, "racket--repl-command: Process racket-command not running". It means that Emacs lost its connection to the Racket TCP/IP backend "server".

**Comment by pmatos** [comment]
Sorry for only coming back to this now. I did as you suggested. All fine. REPL starts, REPL exits and I can start it again but `M-.` on a function generates `Not found. Run current buffer and try again? (y or n)`

Pressing `y` just reloads the buffer on the repl and shows the same question again, over and over. Pressing `n` just breaks the loop and returns me to the buffer but does not find the function definition I originally requested.

**Comment by pmatos** [comment]
Not sure if it matters but this is running `racket 6.11`

**Comment by greghendershott** [comment]
I got some time to review open issues and noticed this was left hanging.

Are you still using racket-mode?  If so, I can try again to figure this out.

In my own use, I sometimes encounter `Not found. Run current buffer and try again? (y or n)`. If it doesn't work after the first `y` retry, then "not found" really means that racket-mode just can't find the definition. 

There's no official "API" provided by Racket for this, beyond [`identifier-binding`](https://docs.racket-lang.org/reference/stxcmp.html#(def._((quote._~23~25kernel)._identifier-binding))). So, racket-mode goes through [various gymnastics](https://github.com/greghendershott/racket-mode/blob/master/defn.rkt) trying to find the definition. Some possible failure modes:

- Stale `zo`s. If you ever did a `raco make` or `raco setup`, you need to do it again, or `rm compiled/`. (This usually fixes it, for me.)

- `identifier-binding` can handle renames due to `rename-out` provides and due to contracts -- but not both, as [this comment describes](https://github.com/greghendershott/racket-mode/blob/master/defn.rkt#L59-L63). There's not much racket-mode can do about this, AFAICT.

- Some new, other scenario you've discovered?  If you're willing to debug this with me, e.g. provide one or more simplified example source files, and I can repro it, then I'll make time to investigate.

**Comment by pmatos** [comment]
@greghendershott Still using racket-mode every day. Let me try to reproduce this and get back to you.

**Comment by pmatos** [comment]
OK, I reproduced this with ActivityLog2.

1. Clone activitylog2: `https://github.com/alex-hhh/ActivityLog2.git`
2. Visit `rkt/trend-charts/view-trends.rkt`
3. Go to line 342: `(let ((ct (send (new new-trend-chart-dialog%) show-dialog parent)))`
4. With the cursor over the first char `n` of `new-trend-chart-dialog...` I press `Alt-.`
5. I get `Not found. Run the current buffer and try again? (y or n)`
6. No matter how many times I press `y` I keep getting the same thing over and over again.

**Comment by greghendershott** [comment]
Thanks!

I had a tough time getting this to run.

When I <kbd>C-c C-c</kbd> in `view-trends.rkt` I got

```
*** (embedded-version): file-size: cannot get size
  path: /Users/greg/src/racket/ActivityLog2/rkt/trend-charts/./version.txt
  system error: No such file or directory; errno=2
```

and then it seemed to hang.

I fixed that (I think?) by:

```diff
modified   rkt/utilities.rkt
@@ -61,7 +61,11 @@
           racket/base
           racket/file
           racket/format
-          racket/string))
+          racket/string
+          racket/runtime-path))
+
+(begin-for-syntax
+  (define-runtime-path version.txt "../version.txt"))
 
 ;; Evaluate to a string containing the version as stored in the ../version.txt
 ;; file.
@@ -74,7 +78,7 @@
                         "no version")))
                     ;; NOTE: version.txt is loaded from the current directory
                     ;; of the compilation, which is the toplevel directory.
-                    (let ((version (string-trim (file->string "./version.txt" #:mode 'text))))
+                    (let ((version (string-trim (file->string version.txt #:mode 'text))))
                       version)))]))
```

No more error message but it _still_ seemed to hang.

Then I realized I had `racket-error-context` set to `'high`. I set it to `'medium`. Now I get a REPL prompt... after about 30 seconds.

---

So now I try <kbd>M-.</kbd> and I get "command process timeout". Which doesn't surprise me, if things are this slow, as I have `racket-command-timeout` set to the default `10` seconds.

So I `(setq racket-command-timeout 60)` and try again. Still get timeout message. How about `120`? Still a timeout.

---

Thoughts so far:

- I'm surprised you don't get the command timeout message instead of another y/n prompt. Maybe this is because you have a `compiled/view-trends.zo`?

- `identifier-binding` tells racket-mode the source file the definition comes from. To locate it within the file, racket-mode has to `expand` the source file, then walk the syntax. So, if something takes 30 seconds to run without a compiled `zo`, it's going to take at least 30 seconds for racket-mode to do its stuff.  This may be a scenario that racket-mode just can't handle well.

- If you answer yes to one "try to run?" prompt, and it doesn't work?  It's not going to work on a 3rd or more attempt. But I shouldn't assume that's obvious to the user. Maybe it shouldn't even prompt again, and say something ~= "sorry, this seems to be a situation where we can't find the definition quickly or at all".

But I'll keep thinking about this some more....

**Comment by greghendershott** [comment]
@alex-hhh As for the rest of it, your `building.md` does talk about the value of compiled zos, and also warns how slow `errortrace` will be. So that's as-expected.

**Comment by greghendershott** [comment]
So I did what `building.md` suggests: `raco make run.rkt rkt/main.rkt`.

That took quite awhile to finish!

After it did, when I go to `view-trends.rkt` line 342, cursor on `new-trend-chart-dialog%`, and <kbd>M-.</kbd> -- it works within a second or two and I'm on the definition at line 124.

So:

- racket-mode's expand is benefiting from the compiled zo, which is great.

- I need to figure out why I can't reproduce @pmatos's  [recipe](https://github.com/greghendershott/racket-mode/issues/288#issuecomment-402434222).

**Comment by alex-hhh** [comment]
Hi @greghendershott, the message is actually printed by my code not by Racket and it is harmless: there is a `with-handlers` that handles the exception, but also prints it out.  However, I agree that it is confusing so I fixed it.

I don't run into this message when working on individual racket files inside the project, as there is always a compiled utilities.zo file available and this code runs only in the compilation step, so relative paths don't matter at run time.  You and @pmatos may be the only two other people who tried to compile and run this code, thanks for taking the time to track down problems.

Regarding the bug that was reported here, `M-.` on the `trend-chart-dialog%` works fine for me, so I cannot reproduce either.  It is however slow (1-2 seconds, as you noticed).

As for being slow when the code is not compiled, there are 37000 lines of Racket code (excluding comments and white space)

**Comment by greghendershott** [comment]
@alex-hhh Thanks for the information!

Yes I definitely didn't intend it as a critique of your project. I've worked on some Racket projects where you definitely need compiled `zo`s -- because of so much code and/or using Typed Racket or other non-trivial expansion. But I haven't so much, lately, and it caught me off guard -- especially after seeing that error message that (seemed to) cause a "hang".  Programming is fun :smile:

**Comment by pmatos** [comment]
OK, now it seems we are in the realm of... something is wrong with my setup. 
I tried again with `raco make` which I was not doing and now I managed to get it to work.

However, I have seen strange cases where it doesn't work so I tried again another identifier `after-move` and this time it didn't work. Take a look: https://asciinema.org/a/qmPTJEFifKcdk5HfVkiHJk8Th

This is with racket 6.12 and after `raco make run.rkt rkt/main.rkt`

**Comment by pmatos** [comment]
Just to clarify what's going on here, I press `alt-.` and it tells me `Not found.... (y or n)`. When I press `y` I get the same thing straight away over and over again without jumping to the point of definition.

**Comment by alex-hhh** [comment]
I played with the `M-.` command in racket mode a bit mode (I haven't really used it until now), and I noticed that it is only able to find things that are `define`d at top level: it works fine for going to a class definition or a function definition, but if I point it to a method name, even one that is defined in the same class, it will enter this "Not found. Run current buffer and try again?" loop.

I do not know what is involved in finding definitions, but  perhaps it should just give up if it evaluated the current buffer and could not find the definition?

**Comment by greghendershott** [comment]
@pmatos / @alex-hhh 
>> `after-move`
>  it is only able to find things that are defined at top level: it works fine for going to a class definition or a function definition

Yes. Its first step is to use Racket's [`identifier-binding`](https://docs.racket-lang.org/reference/stxcmp.html#(def._((quote._%7E23%7E25kernel)._identifier-binding))) to determine the file(s) that define and provide the identifier. If that returns `#f` or `'lexical`, it gives up.

>  but if I point it to a method name, even one that is defined in the same class,

In that case `identifier-binding` returns `#f`.

And also e.g. if you point to `inner` in:

```
(define (outer)
  (define (inner) ___) ___)
```

I've worked on this so long, it's "obvious" to me `identifier-binding` won't work here. So I didn't even think to include it in my list of "failure modes" above.  But it's reasonable to expect it to work.

_Maybe_ when `identifier-binding` has nothing racket-mode could, up in Emacs Lisp, try to find the definition in the current source file textually using regexps?  I mean, that's often what I would do manually if <kbd>M-.</kbd> failed. Why can't racket-mode try this for me?  It won't always succeed, but it often could.

**Comment by greghendershott** [comment]
@alex-hhh 
> perhaps it should just give up if it evaluated the current buffer and could not find the definition?

It's meant to address some common cases:

1. You're editing `foo.rkt` but what's live in the REPL is `bar.rkt`.

2. `foo.rkt` is live both places, but you edited it to add some definitions, and haven't yet _re_-run it (so the _current_ version isn't yet live in the REPL).

The common cure is simply to run `foo.rkt` and try again.

Instead of telling the user to try this, it offers to do it for them.

---

Rather than trying to address both cases with this prompt after <kbd>M-.</kbd> fails, maybe it could handle them differently?

1. Different file. Here it may as well go ahead and prompt to run `foo.rkt` _before_ even attempting to find the definition. And if that fails, just fail.  This should be easy to implement because the Emacs Lisp side already has a way to ask the Racket back-end what filename is live in the REPL.

2. Not current.
    (a) Edited-but-not-saved is easy to detect.
    (b) Edited and already saved is tricky; there's no checksum of the source in Emacs and in the racket-mode "backend" to compare. I can think about this.

**Note:** I think racket-mode must prompt and _offer_ to run for the user -- never go ahead and do it. Because running can have side-effects.

**Comment by alex-hhh** [comment]
I understand why racket-mode has to run the current buffer once if an identifier is not found;  However, if the identifier is still not found after running the buffer once, it should not prompt the user to run it a second time -- this is where it enters a loop.

----

Regarding the textual search, I don't think this is a good idea: as it could create more confusion on complex code bases, where accuracy is needed the most.

Looking at my code base, there are places where a class is passed around to another class, which instantiates an object and invokes methods.   It is not possible to know where is the implementation of those methods, except at run time.   Since the code uses general name (e.g. "restore-from") which are implemented by multiple unrelated classes, there is a big chance that `M-.`  will select the wrong method.

If the user is forced to do a textual search after a failed `M-.` call, they would presumably have a good idea that the textual search might return wrong results.

**Comment by greghendershott** [comment]
Re the prompt part of this. I spent awhile today exploring different approaches. I think I'm ready to push a commit. Overview:

1. The back-end has an existing command, `prompt`, which returns the /path/to/file.rkt of the running file IFF the REPL is waiting at a prompt. I extended this to also return an MD5 checksum of the file.

2. Then in Elisp on the front-end:

```elisp
(defun racket--do-visit-def-or-mod (cmd sym)
  "CMD must be \"def\" or \"mod\". SYM must be `symbolp`."
  (unless (racket--repl-at-prompt-for-our-buffer-p)
    (when (y-or-n-p "Run current buffer first? ")
      (racket--run-and-wait-for-prompt)))
  (pcase (racket--repl-command "%s %s" cmd sym)
    (`(,path ,line ,col)
     (racket--push-loc)
     (find-file path)
     (goto-char (point-min))
     (forward-line (1- line))
     (forward-char col)
     (message "Type M-, to return"))
    (`kernel
     (message "`%s' defined in #%%kernel -- source not available." sym))
    (_
     (message "Not found."))))

(defun racket--repl-at-prompt-for-our-buffer-p ()
  "Is the REPL at a prompt, and evaluated for our buffer contents?"
  (equal (racket--repl-command "prompt")
         (cons (buffer-file-name) (md5 (current-buffer)))))

(defun racket--run-and-wait-for-prompt ()
  "Do `racket-run', then wait for the prompt to match our .rkt file."
  (racket-run)
  (with-timeout (racket-command-timeout
                 (error "Command timeout waiting for racket-run to finish"))
    (while (not (racket--repl-at-prompt-for-our-buffer-p))
      (message "Waiting for prompt...")
      (sit-for 0.1))))
```

As a result: When you <kbd>M-.</kbd>, first it checks whether the REPL has the correct file and it matches what's in the buffer. Only if not, it prompts you to `racket-run' y/n.  In any case, next it tries to find the definition. If it fails, there's no point is asking to run again -- it can confidently just say "Not found", the end.

> Sidebar: I first tried doing this with file modification times instead of MD5 hash. But that's
> - less reliable
> - requires a `save-buffer` _before_ even asking about running (blech)
> - actually more of a headache to code because:
>    - extra state the back-end would need to store (file mod time when last run)
>    - Emacs' interesting `file-attributes` time formats, `(list high-16-bytes, low-16-bytes, microseconds, picosends)`

I won't push this to `master` for awhile. I want to dogfood it for awhile, as well as try to think of flaws. But, so far, it seems to be working really nicely!

**Comment by greghendershott** [comment]
> Regarding the textual search, I don't think this is a good idea: as it could create more confusion on complex code bases, where accuracy is needed the most.

I think you're probably right. Also I didn't even get that far, today.

I'd like to at least give a hint why it wasn't found, in cases such as @pmatos's example.

Maybe the message could be something like the following: `"Not found. racket-mode can't reliably find definitions of class members and local functions; try text search"`.

What do you think? Any suggestions for better message text, or anything?

**Comment by greghendershott** [comment]
Maybe the message should be just `"Not found. C-h racket-visit-definition for more information."` and then I can write a clearer paragraph or two in the doc string.

**Comment by greghendershott** [comment]
More re [the `view-trends.rkt` example](https://github.com/greghendershott/racket-mode/issues/288#issuecomment-402525017):

> After [doing `raco make`], when I go to view-trends.rkt line 342, cursor on new-trend-chart-dialog%, and M-. -- it works within a second or two and I'm on the definition at line 124.
> racket-mode's expand is benefiting from the compiled zo, which is great.

The fact that this can take seconds (as opposed to milliseconds) is unfortunate. But I don't see how to make it faster, while still being reliable.

The reason it is slow is due to the need to `expand` the syntax from the `view-trends.rkt` file -- `expand` dominates the time here (not walking the resulting syntax to find the definition).

After a `raco make`, the `view-trends.rkt` file's `require` of _other_ files _does_ benefit from the `zo`s produced by `raco  make` -- the `zo`s make it take a few seconds instead of tens of seconds!

But the `view-trends.rkt` file itself must still be `expand`ed from source -- and that takes a few seconds instead of milliseconds.

---

- If `expand` is (for some files) so slow, can we avoid it?  I don't see how. It's quite common for definitions to be created by "definer" macros, and these definitions can only be discovered in fully-expanded code.

- If we can't avoid it, could we at least shift the cost earlier than <kbd>M-.</kbd>? Could racket-mode generate expansions ahead of time -- storing them under `compiled/` in (say) `.expanded` files, or in a new `expanded/` subdir?  In theory, sure. But:

    - This seems like something that should be done by a `raco` tool that people explicitly opt in to, much like `raco make`. (Note there exists a `raco expand` tool -- however its output isn't `syntax` with the source location information that we need, it's plain text `syntax->datum` of the expansion.)

    - The file that must be expanded is the "target" of the find-definition -- where the definition is. Even if you were to `raco new-tool` all the files in _your_ project, that won't speed up finding definitions in _other_ projects.

TL;DR: I don't see an actionable way to make this faster on source files that are inherently slow to be fully-expanded.

**Comment by alex-hhh** [comment]
I don't think that racket-mode can do anything to make the lookup faster, as you noticed, racket itself is slow when expanding a module, and this is most visible on large code bases.

Another approach used by Emacs (and vi) is the TAGS file approach.  It would be relatively simple to write a parser which finds all things which are defined at top level and constructs a TAGS file -- in fact I suspect the etags program supplied with Emacs already does that.  Since the tags program does not have to know about what is being defined, only that something is being defined, the program can be very fast.   Also the TAGS file format allows for some searching for  tag, which means that editing a file will not immediately invalidate the TAGS file.

**Comment by greghendershott** [comment]
I'm going to mark this as closed. I think I've addressed all the actionable items related to it, in  commits over the last couple weeks, which I just merged these to `master`.

- commit 12513b0 racket-visit-{definition module}: Minimize "run?" prompts
- commit 5c989ab Various changes to find-definition and friends
- commit 1dd6860 Add a cache of expanded syntax.
